File: connectors-editors/basecamp-editor/src/main/java/com/taskadapter/connector/basecamp/editor/BasecampEditorFactory.java
package com.taskadapter.connector.basecamp.editor;

import com.google.common.base.Strings;
import com.taskadapter.common.ui.FieldMapping;
import com.taskadapter.connector.basecamp.BasecampConfig;
import com.taskadapter.connector.basecamp.BasecampConnector;
import com.taskadapter.connector.basecamp.BasecampUtils;
import com.taskadapter.connector.basecamp.BasecampValidator;
import com.taskadapter.connector.basecamp.beans.BasecampProject;
import com.taskadapter.connector.basecamp.beans.TodoList;
import com.taskadapter.connector.basecamp.transport.BaseCommunicator;
import com.taskadapter.connector.basecamp.transport.ObjectAPI;
import com.taskadapter.connector.basecamp.transport.ObjectAPIFactory;
import com.taskadapter.connector.definition.WebConnectorSetup;
import com.taskadapter.connector.definition.exceptions.BadConfigException;
import com.taskadapter.connector.definition.exceptions.ConnectorException;
import com.taskadapter.connector.definition.exceptions.ProjectNotSetException;
import com.taskadapter.model.NamedKeyedObject;
import com.taskadapter.model.NamedKeyedObjectImpl;
import com.taskadapter.web.ConnectorSetupPanel;
import com.taskadapter.web.DroppingNotSupportedException;
import com.taskadapter.web.PluginEditorFactory;
import com.taskadapter.web.callbacks.DataProvider;
import com.taskadapter.web.configeditor.EditorUtil;
import com.taskadapter.web.configeditor.server.ServerPanelWithPasswordAndAPIKey;
import com.taskadapter.web.data.Messages;
import com.taskadapter.web.service.Sandbox;
import com.taskadapter.web.uiapi.DefaultSavableComponent;
import com.taskadapter.web.uiapi.SavableComponent;
import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.checkbox.Checkbox;
import com.vaadin.flow.component.formlayout.FormLayout;
import com.vaadin.flow.component.html.Label;
import com.vaadin.flow.component.textfield.TextField;
import com.vaadin.flow.data.binder.Binder;
import com.vaadin.flow.data.binder.ValidationException;

import java.util.ArrayList;
import java.util.List;

public class BasecampEditorFactory implements PluginEditorFactory<BasecampConfig, WebConnectorSetup> {
    private static final String BUNDLE_NAME = "com.taskadapter.connector.basecamp.editor.messages";
    private static final Messages MESSAGES = new Messages(BUNDLE_NAME);
    private static final ObjectAPIFactory factory = new ObjectAPIFactory(new BaseCommunicator());
    private static final BasecampErrorFormatter formatter = new BasecampErrorFormatter();

    @Override
    public SavableComponent getMiniPanelContents(Sandbox sandbox, BasecampConfig config, WebConnectorSetup setup) {
        Binder<BasecampConfig> binder = new Binder<>(BasecampConfig.class);

        Component projectPanel = createProjectPanel(binder, config, setup);
        FormLayout layout = new FormLayout();
        layout.setResponsiveSteps(new FormLayout.ResponsiveStep("50em", 1),
                new FormLayout.ResponsiveStep("50em", 2));

        layout.add(projectPanel);
        binder.readBean(config);
        return new DefaultSavableComponent(layout, () -> {
            try {
                binder.writeBean(config);
                return true;
            } catch (ValidationException e) {
                e.printStackTrace();
                return false;
            }
        });
    }

    @Override
    public boolean isWebConnector() {
        return true;
    }

    @Override
    public ConnectorSetupPanel getEditSetupPanel(Sandbox sandbox, WebConnectorSetup setup) {
        return new ServerPanelWithPasswordAndAPIKey(BasecampConnector.ID, BasecampConnector.ID, setup);
    }

    @Override
    public WebConnectorSetup createDefaultSetup(Sandbox sandbox) {
        return WebConnectorSetup.apply(BasecampConnector.ID,
                "My Basecamp 2", ObjectAPI.BASECAMP_URL, "", "", false, "");
    }

    @Override
    public List<BadConfigException> validateForSave(BasecampConfig config, WebConnectorSetup setup, List<FieldMapping<?>> fieldMappings) {
        List<BadConfigException> list = new ArrayList<>();

        if (Strings.isNullOrEmpty(config.getProjectKey())) {
            list.add(new ProjectNotSetException());
        }
        return list;
    }

    @Override
    public List<BadConfigException> validateForLoad(BasecampConfig config, WebConnectorSetup setup) {
        return BasecampValidator.validateConfig(config);
    }

    @Override
    public void validateForDropInLoad(BasecampConfig config) throws DroppingNotSupportedException {
        throw DroppingNotSupportedException.INSTANCE;
    }

    @Override
    public String describeSourceLocation(BasecampConfig config, WebConnectorSetup setup) {
        return setup.getHost();
    }

    @Override
    public String describeDestinationLocation(BasecampConfig config, WebConnectorSetup setup) {
        return describeSourceLocation(config, setup);
    }

    @Override
    public Messages fieldNames() {
        return MESSAGES;
    }

    @Override
    public String formatError(Throwable e) {
        return formatter.formatError(e);
    }


    private Component createProjectPanel(Binder<BasecampConfig> binder, BasecampConfig config, WebConnectorSetup setup) {

        Label accountIdLabel = new Label("Account Id");
        TextField accountIdField = EditorUtil.textInput(binder, "accountId");

        Label projectLabel = new Label("Project key");
        TextField projectField = EditorUtil.textInput(binder, "projectKey");
        Button projectInfoButton = EditorUtil.createButton("Info", "View the project info",
                event -> ShowInfoElement.loadProject(config, setup, formatter, factory)
        );
        DataProvider<List<? extends NamedKeyedObject>> projectProvider = new DataProvider<List<? extends NamedKeyedObject>>() {
            @Override
            public List<? extends NamedKeyedObject> loadData() throws ConnectorException {
                List<BasecampProject> basecampProjects = BasecampUtils.loadProjects(factory, config, setup);
                List<NamedKeyedObject> objects = new ArrayList<>();
                for (BasecampProject project : basecampProjects) {
                    objects.add(new NamedKeyedObjectImpl(project.getKey(), project.getName()));
                }
                return objects;
            }
        };

        Button showProjectsButton = EditorUtil.createLookupButton(
                "...",
                "Show list of available projects on the server.",
                "Select a project",
                projectProvider,
                formatter,
                namedKeyedObject -> {
                    projectField.setValue(namedKeyedObject.getKey());
                    return null;
                }
        );

        Label todoListKeyLabel = new Label("Todo list key");
        TextField todoListKeyField = EditorUtil.textInput(binder, "todoKey");
        Button todoListInfoButton = EditorUtil.createButton("Info", "View the todo list info",
                event -> ShowInfoElement.showTodoListInfo(config, setup, formatter, factory)
        );

        DataProvider<List<? extends NamedKeyedObject>> todoListsProvider = () -> {
            List<TodoList> todoLists = BasecampUtils.loadTodoLists(factory, config, setup);
            List<NamedKeyedObject> objects = new ArrayList<>();
            for (TodoList todoList : todoLists) {
                objects.add(new NamedKeyedObjectImpl(todoList.getKey(), todoList.getName()));
            }
            return objects;
        };

        Button showTodoListsButton = EditorUtil.createLookupButton(
                "...",
                "Show Todo Lists",
                "Select a Todo list",
                todoListsProvider,
                formatter,
                namedKeyedObject -> {
                    todoListKeyField.setValue(namedKeyedObject.getKey());
                    return null;
                }
        );

        Checkbox findUserByName = EditorUtil.checkbox("Find users based on assignee's name",
                "This option can be useful when you need to export a new MSP project file to Redmine/JIRA/MantisBT/....\n"
                        + "Task Adapter can load the system's users by resource names specified in the MSP file\n"
                        + "and assign the new tasks to them.\n"
                        + "Note: this operation usually requires 'Admin' permission in the system.",
                binder, "findUserByName");

        Checkbox loadCompletedItemsCheckbox = EditorUtil.checkbox("Load completed items",
                "",
                binder, "loadCompletedTodos");

        FormLayout layout = new FormLayout();
        layout.setResponsiveSteps(new FormLayout.ResponsiveStep("20em", 1),
                new FormLayout.ResponsiveStep("60em", 2),
                new FormLayout.ResponsiveStep("20em", 3),
                new FormLayout.ResponsiveStep("20em", 4));

        layout.add(accountIdLabel);
        layout.add(accountIdField, 3);

        layout.add(projectLabel, projectField, projectInfoButton, showProjectsButton);

        layout.add(todoListKeyLabel, todoListKeyField, todoListInfoButton, showTodoListsButton);

        layout.add(loadCompletedItemsCheckbox, 4);
        layout.add(findUserByName, 4);

        return layout;
    }

}

File: connectors-editors/basecamp-editor/src/main/scala/com/taskadapter/connector/basecamp/editor/BasecampEditorFactory.scala
package com.taskadapter.connector.basecamp.editor

import java.util
import com.taskadapter.connector.basecamp.transport.{BaseCommunicator, ObjectAPI, ObjectAPIFactory}
import com.taskadapter.connector.basecamp.{BasecampConfig, BasecampConnector, BasecampUtils, BasecampValidator}
import com.taskadapter.connector.definition.exceptions.{BadConfigException, ConnectorException, ProjectNotSetException}
import com.taskadapter.connector.definition.{FieldMapping, WebConnectorSetup}
import com.taskadapter.model.{NamedKeyedObject, NamedKeyedObjectImpl}
import com.taskadapter.vaadin14shim.GridLayout
import com.taskadapter.web.callbacks.DataProvider
import com.taskadapter.web.configeditor.EditorUtil.{propertyInput, textInput}
import com.taskadapter.web.configeditor.server.ServerPanelFactory
import com.taskadapter.web.configeditor.{EditorUtil, Editors}
import com.taskadapter.web.data.Messages
import com.taskadapter.web.service.Sandbox
import com.taskadapter.web.ui.Grids.addTo
import com.taskadapter.web.{ConnectorSetupPanel, DroppingNotSupportedException, PluginEditorFactory}
import com.vaadin.data.util.MethodProperty
import com.vaadin.ui.{Alignment, CheckBox, HasComponents, Label, Panel}

import scala.collection.JavaConverters._
import scala.collection.Seq

class BasecampEditorFactory extends PluginEditorFactory[BasecampConfig, WebConnectorSetup] {
  private val BUNDLE_NAME = "com.taskadapter.connector.basecamp.editor.messages"
  private val MESSAGES = new Messages(BUNDLE_NAME)
  private val factory = new ObjectAPIFactory(new BaseCommunicator)
  private val formatter = new BasecampErrorFormatter

  override def isWebConnector = true

  override def getMiniPanelContents(sandbox: Sandbox, config: BasecampConfig, setup: WebConnectorSetup): HasComponents = {
    val projectPanel = createProjectPanel(config, setup)
    val grid = new GridLayout
    grid.setColumns(2)
    grid.setMargin(true)
    grid.setSpacing(true)
    grid.add(projectPanel)
    grid
  }

  override def getEditSetupPanel(sandbox: Sandbox, setup: WebConnectorSetup): ConnectorSetupPanel = ServerPanelFactory.withApiKeyAndLoginPassword(BasecampConnector.ID, BasecampConnector.ID, setup)

  override def createDefaultSetup(sandbox: Sandbox) = new WebConnectorSetup(BasecampConnector.ID, Option.empty, "My Basecamp 2", ObjectAPI.BASECAMP_URL, "", "", false, "")

  private def createProjectPanel(config: BasecampConfig, setup: WebConnectorSetup) = {
    val projectPanel = new Panel("Project")
    val grid = new GridLayout
    grid.setColumns(4)
    grid.setMargin(true)
    grid.setSpacing(true)
    projectPanel.setContent(grid)
    addAccountIdRow(config, grid)
    addProjectRow(config, setup, grid)
    addTodoKeyRow(config, setup, grid)
    addCompletedCheckboxRow(config, grid)
    addFindUsersCheckboxRow(config, grid)
    projectPanel
  }

  private def addFindUsersCheckboxRow(config: BasecampConfig, grid: GridLayout) = {
    grid.add(Editors.createFindUsersElement(new MethodProperty[java.lang.Boolean](config, "findUserByName")))
    grid.add(new Label(""))
    grid.add(new Label(""))
    grid.add(new Label(""))
  }

  private def addCompletedCheckboxRow(config: BasecampConfig, grid: GridLayout) = {
    val loadCompletedTasksCheckbox = new CheckBox("Load completed items")
    loadCompletedTasksCheckbox.setPropertyDataSource(new MethodProperty[String](config, "loadCompletedTodos"))
    grid.add(loadCompletedTasksCheckbox)
    grid.add(new Label(""))
    grid.add(new Label(""))
    grid.add(new Label(""))
  }

  private def addAccountIdRow(config: BasecampConfig, grid: GridLayout) = {
    val accountIdLabel = new Label("Account Id:")
    grid.add(accountIdLabel)
    val accountIdField = propertyInput(config, "accountId")
    grid.add(accountIdField)
    grid.add(new Label(""))
    grid.add(new Label(""))
  }

  private def addProjectRow(config: BasecampConfig, setup: WebConnectorSetup, grid: GridLayout) = {
    val projectKeyLabel = new Label("Project key:")
    addTo(grid, Alignment.MIDDLE_LEFT, projectKeyLabel)
    val projectKeyProperty = new MethodProperty[String](config, "projectKey")
    addTo(grid, Alignment.MIDDLE_LEFT, textInput(projectKeyProperty))
    val infoButton = EditorUtil.createButton("Info", "View the project info",
      _ => ShowInfoElement.loadProject(config, setup, formatter, factory)
    )

    addTo(grid, Alignment.MIDDLE_CENTER, infoButton)
    val projectProvider = new DataProvider[util.List[_ <: NamedKeyedObject]]() {
      @throws[ConnectorException]
      override def loadData: util.List[_ <: NamedKeyedObject] = {
        val basecampProjects = BasecampUtils.loadProjects(factory, config, setup)
        val objects = new util.ArrayList[NamedKeyedObject]
        import scala.collection.JavaConversions._
        for (project <- basecampProjects) {
          objects.add(new NamedKeyedObjectImpl(project.getKey, project.getName))
        }
        objects
      }
    }
    val showProjectsButton = EditorUtil.createLookupButton("...",
      "Show list of available projects on the server.",
      "Select project",
      "List of projects on the server",
      projectProvider,
      formatter,
      (namedKeyedObject: NamedKeyedObject) => {
        def foo(namedKeyedObject: NamedKeyedObject) = {
          projectKeyProperty.setValue(namedKeyedObject.getKey)
          null
        }

        foo(namedKeyedObject)
      })
    addTo(grid, Alignment.MIDDLE_CENTER, showProjectsButton)
  }

  private def addTodoKeyRow(config: BasecampConfig, setup: WebConnectorSetup, grid: GridLayout) = {
    val todoListKey = new Label("Todo list key:")
    addTo(grid, Alignment.MIDDLE_LEFT, todoListKey)
    val todoKeyProperty = new MethodProperty[String](config, "todoKey")
    addTo(grid, Alignment.MIDDLE_LEFT, textInput(todoKeyProperty))
    val infoButton = EditorUtil.createButton("Info", "View the todo list info",
      _ => ShowInfoElement.showTodoListInfo(config, setup, formatter, factory)
    )
    addTo(grid, Alignment.MIDDLE_CENTER, infoButton)
    val todoListsProvider: DataProvider[java.util.List[_ <: NamedKeyedObject]] = () => {
      val todoLists = BasecampUtils.loadTodoLists(factory, config, setup)
      todoLists.asScala.map(todoList => NamedKeyedObjectImpl(todoList.getKey, todoList.getName)).asJava.asInstanceOf[java.util.List[NamedKeyedObject]]
    }
    val showTodoListsButton = EditorUtil.createLookupButton("...", "Show Todo Lists",
      "Select a Todo list",
      "Todo lists on the server",
      todoListsProvider,
      formatter,
      (namedKeyedObject: NamedKeyedObject) => {
        def foo(namedKeyedObject: NamedKeyedObject) = {
          todoKeyProperty.setValue(namedKeyedObject.getKey)
          null
        }

        foo(namedKeyedObject)
      })
    addTo(grid, Alignment.MIDDLE_CENTER, showTodoListsButton)
  }

  override def validateForSave(config: BasecampConfig, setup: WebConnectorSetup, fieldMappings: Seq[FieldMapping[_]]): Seq[BadConfigException] = {
    if (config.getProjectKey == null || config.getProjectKey.isEmpty) {
      return Seq(new ProjectNotSetException)
    }
    Seq()
  }

  override def validateForLoad(config: BasecampConfig, setup: WebConnectorSetup): Seq[BadConfigException] = BasecampValidator.validateConfig(config)

  override def describeSourceLocation(config: BasecampConfig, setup: WebConnectorSetup): String = setup.host

  override def describeDestinationLocation(config: BasecampConfig, setup: WebConnectorSetup): String = describeSourceLocation(config, setup)

  override def fieldNames: Messages = MESSAGES

  override def formatError(e: Throwable): String = formatter.formatError(e)

  @throws[DroppingNotSupportedException]
  override def validateForDropInLoad(config: BasecampConfig) = throw DroppingNotSupportedException.INSTANCE
}